### ISSUE TITLE
RFE: Remove Provider Specific details from e2e tests

### ISSUE THREAD

**Issue by timothysc** [kubernetes/kubernetes#70194]
**Is this a FEATURE REQUEST?**:
The e2e tests have details baked in that have existed since the early days around providers.  Recent refactoring has exposed brittleness in how we invoke the tests and TBH provider specific behavioral tests that depend on non-standard APIs should be moved outside of tree.

/kind feature
/sig testing 
/cc @neolit123 @pohly @BenTheElder @kubernetes/cncf-conformance-wg @kubernetes/sig-testing @kubernetes/k8s-infra-team

**Comment by hogepodge** [comment]
sig-cloud-provider can notify individual providers and help assign work on this enhancement

**Comment by andrewsykim** [comment]
@timothysc do we have any details on what tests specifically need changing or are there too many to list?

**Comment by andrewsykim** [comment]
nvm, I did a quick scan and it's everywhere 😬

**Comment by pohly** [comment]
Is the goal to not have any dependency (whether it is direct or indirect) on provider-specific functionality in Kubernetes? The idea behind the `vendor` interface was that tests can still be developed in-tree and different cloud providers can share the same test logic. That doesn't have to be in Kubernetes itself, though; "shared tests" could also be its own repo.

Another concern was raised in the CSI working group, independent of this PR here: when removing tests that depend on vendor-specific functionality, Kubernetes CI by itself might exercise less code paths in Kubernetes core and thus miss regressions that currently get caught.

**Comment by neolit123** [comment]
my vote would go for no provider implementations in k/k, only interfaces or a framework that can be implemented in third party repos.
in terms of CI, test jobs can always pull code from third party repos and build + run tests so that k8s has good coverage.

**Comment by andrewsykim** [comment]
There's a muddled boundary between what is "provider specific" and what tests are "generic" but only pass on certain providers (like [this test](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apps/rc.go#L51-L57)). The latter becomes harder to reason with because we need to decouple the provider specific parts in Kubernetes itself (an effort we are working on now) before we we can decouple it in the tests. There may be intermediate steps we have to take to get us closer to that point but I agree with @neolit123 that an ideal end result is that there are no provider tests in k/k and each provider can run the core tests suites in k/k and their own provider specific test suits (which lives in their own provider repos).

**Comment by andrewsykim** [comment]
@pohly to answer your question directly, yes the goal is the get rid of all cloud provider depenencies in `k8s.io/kubernetes`. Though there will be a transitional period where we'll vendor `k8s.io/cloud-provider-<provider>` instead of `k8s.io/kubernetes/pkg/cloudprovider/providers/<provider>` so all the dependencies of a cloud provider can be easily added/removed from k/k.

**Comment by BenTheElder** [comment]
+1 to what @neolit123 said, though I will add that presumably these tests are necessarily coupled to the provider implementations, and imo should be versioned with them, so I think probably they should move out together, provider and provider tests.

> Another concern was raised in the CSI working group, independent of this PR here: when removing tests that depend on vendor-specific functionality, Kubernetes CI by itself might exercise less code paths in Kubernetes core and thus miss regressions that currently get caught.

I don't think this is acceptable. The CSI interface should be fully test-able without vendor specific details in the tests and vendor specific testing should happen in the driver repos. That's sort of the point of having an interface imo... things should be able to target the _interface_, including the tests. We should be able to use various "local" drivers or generic ones like say NFS to test within the repo, and test other drivers elsewhere..

To the same point, we don't test all CRI in presubmit, or all cloud providers or... we'd never merge anything with that matrix.

We can always put vendor specific CSI tests in release blocking CI without being in the core repo anyhow if we really need to. 

More generally we should not catch literally all regressions via pre-submit-test-all-the-things in kubernetes/kubernetes. We already cannot. If the tests aren't in k/k presubmit, it's not particularly important from a testing POV where the code lives, probably it should live close to the vendor specific code it tests. 
We have post-submit and continuous testing for this which can use any number of repos.


cc @davidz627

**Comment by pohly** [comment]
Benjamin Elder <notifications@github.com> writes:
> +1 to what @neolit123 said, though I will add that presumably these
> tests are necessarily coupled to the provider implementations, and imo
> should be versioned with them, so I think probably they should move
> out together, provider and provider tests.

Some tests aren't specific to *one* provider, but depend on
functionality provided only by *some* providers via provider-specific
APIs. If those kind of tests aren't welcome in Kubernetes itself
anymore, then we'll need this extra "shared tests repo", because they
can't (or rather, shouldn't) be moved into the repos of each vendor
either.

>> Another concern was raised in the CSI working group, independent of
>> this PR here: when removing tests that depend on vendor-specific
>> functionality, Kubernetes CI by itself might exercise less code paths
>> in Kubernetes core and thus miss regressions that currently get
>> caught.
>
> I don't think this is acceptable. The CSI interface should be fully
> test-able without vendor specific details in the tests and vendor
> specific testing should happen in the driver repos. That's sort of the
> point of having an interface imo... things should be able to target
> the _interface_, including the tests. We should be able to use various
> "local" drivers or generic ones like say NFS to test within the repo,
> and test other drivers elsewhere..

Something based on NFS is indeed being discussed at the moment, but it
is ready yet. There are ways to mitigate this issue, but not always is
mocking close enough to the real thing.

> More generally we should not catch literally all regressions via
> pre-submit-test-all-the-things in kubernetes/kubernetes. We already
> cannot. If the tests aren't in k/k presubmit, it's not particularly
> important from a testing POV where the code lives, probably it should
> live close to the vendor specific code it tests.  We have post-submit
> and continuous testing for this which can use any number of repos.

That sounds like a useful approach.

**Comment by BenTheElder** [comment]
> If those kind of tests aren't welcome in Kubernetes itself
anymore, then we'll need this extra "shared tests repo", because they
can't (or rather, shouldn't) be moved into the repos of each vendor
either.

That seems unfortunate, ideally that should still be behind a k8s style API / interface. I'm not sure that those are necessarily unwelcome so much as something we should avoid. I'd hope we aim for tests to generally be of high enough quality to move towards conformance unless they are:
- testing provider details (which belongs to the provider to manage in the long term, not k/k...)
- testing performance
- intentionally disruptive

EG that private registry test, surely we could just run a local private registry in the test cluster instead of depending on GCE / GKE?

> Something based on NFS is indeed being discussed at the moment, but it
is ready yet. There are ways to mitigate this issue, but not always is
mocking close enough to the real thing.

Well of course, but we should be able to cover the interface surface / paths without any particular vendor ...?
We'd of course not be covering the vendor implementations, but we should cover those where they are implemented, which is presumably out of tree ...?

**Comment by neolit123** [comment]
> Some tests aren't specific to *one* provider, but depend on
functionality provided only by *some* providers via provider-specific
APIs. If those kind of tests aren't welcome in Kubernetes itself
anymore, then we'll need this extra "shared tests repo", because they
can't (or rather, shouldn't) be moved into the repos of each vendor
either.

could you provide an example of such a test?
if a single test is importing bits from different c-provider SDK/API it feels as if such a test should be split into multiple tests to avoid the scenario.

such tests should not reside in k/k.

**Comment by davidz627** [comment]
/cc @saad-ali @msau42 @mkimuram 

For storage:
This is already on our radar. That being said, this would be a huge change and I believe we have a multi-stage plan to address this.
In the near term @mkimuram is doing some refactoring of the tests which helps clarify the distinction between providers (drivers) and test suites (run for each provider).
In the slightly longer term I believe @pohly and @msau42  have expressed interest in refactoring the tests even further to decouple test suites from the driver so that the test suites can be easily consumed from an outside repository (providers)
In the much longer term we plan on migrating all in-tree drivers to CSI (https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-migration.md) and at the end of that process should be able to delete all in-tree storage provider code. This is necessarily coupled with he maturity of the respective CSI Drivers and therefore we can also remove those drivers from the e2e tests. Each provider can then target the k8s/k8s test suites with their own drivers in their third party repositories.

Below is from @msau42:
yes, we're taking baby steps towards reaching that end goal, but until we have the appropriate cloud-provider specific testing infrastructure in place, we can't just remove the tests now. also to the other point about why can't we just test mock drivers in-tree mock drivers don't give us the same level of confidence as a real production driver. mocks are great for unit or integration testing, but we want to test real end-user behavior as much as possible. if there's a way to make an out-of-tree job still release blocking for kubernetes (and it sounds like there is), then i agree that this is a good way forward

**Comment by andrewsykim** [comment]
cc @d-nishi

**Comment by BenTheElder** [comment]
> Also to the other point about why can't we just test mock drivers in-tree mock drivers don't give us the same level of confidence as a real production driver. mocks are great for unit or integration testing, but we want to test real end-user behavior as much as possible. if there's a way to make an out-of-tree job still release blocking for kubernetes (and it sounds like there is), then i agree that this is a good way forward

Not suggesting just "mock" drivers, though those would be great too, but "actual" drivers for say, NFS, which we can run using only what is in tree, or even using other "actual" drivers but making _the test logic_ (IE e2e.test) only depend on EG the CSI, w/o vendor details in the tests.

We can also have other tests block releases (EG some CI job integrating k/k and a vendor driver..), but we really ought to get vendor details out of tree so we can patch those independently of k/k and vice versa, ... that includes the tests currently.

Obviously this is a **long**-term thing, and there are already some efforts working on the non-test code, but we probably should do something similar for the tests _eventually_.

More concretely: The providers moving out of vendor tend to need to depend on k/k there, so the test code cannot in turn depend on the provider code or we have a cycle...

**Comment by msau42** [comment]
No single driver, except maybe a mock driver, can test the full functionality of the CSI interface.  For example, NFS can test mount/unmount, but in order to test attach/detach, you need a cloud-provider volume plugin.  Same goes for other storage features like snapshots and topology.  For storage, at least, we'll need a combination of "common" drivers like nfs along with cloud provider specific plugins to be able to get complete coverage.  I do agree though, that we should work towards moving all provider-specific testing out of tree, but we may still need to aggregate results from in-tree and out-of-tree in order to get the full picture.

**Comment by BenTheElder** [comment]
:+1: Also discussed this offline with @davidz627, I think mostly we had some terminology mismatch. All of the work coming out of storage for this looks great :100:
I think we'll be in a much better place within a few releases.. we (David and I) at least plan to follow up ~beginning of the next cycle about getting specifically some more testing targeting CSI into release-blocking.

**Comment by pohly** [comment]
"Lubomir I. Ivanov" <notifications@github.com> writes:
>> Some tests aren't specific to *one* provider, but depend on
> functionality provided only by *some* providers via provider-specific
> APIs. If those kind of tests aren't welcome in Kubernetes itself
> anymore, then we'll need this extra "shared tests repo", because they
> can't (or rather, shouldn't) be moved into the repos of each vendor
> either.
>
> could you provide an example of such a test?

You can have a look at
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/provider.go
and its ProviderInterface. It has methods that are typically
implemented differently by at least two cloud providers and used by
shared tests.

Let's pick something that isn't for storage, because we've already
discussed that and I want to avoid the impression that this only a
problem for SIG-Storage:
- https://github.com/kubernetes/kubernetes/blob/022c05c1418a75bec494e70fee74b0f4bbf1351b/test/e2e/framework/provider.go#L85-L87
  defines ResizeGroup, GetGroupNodes, GroupSize
- this gets implemented for example in:
  https://github.com/kubernetes/kubernetes/blob/022c05c1418a75bec494e70fee74b0f4bbf1351b/test/e2e/framework/providers/aws/aws.go#L49-L64
  https://github.com/kubernetes/kubernetes/blob/022c05c1418a75bec494e70fee74b0f4bbf1351b/test/e2e/framework/providers/gce/gce.go#L119-L155
  https://github.com/kubernetes/kubernetes/blob/022c05c1418a75bec494e70fee74b0f4bbf1351b/test/e2e/framework/providers/kubemark/kubemark.go#L51-L57
- and used here (indirectly):
  https://github.com/kubernetes/kubernetes/blob/022c05c1418a75bec494e70fee74b0f4bbf1351b/test/e2e/lifecycle/resize_nodes.go#L110-L137
  https://github.com/kubernetes/kubernetes/blob/022c05c1418a75bec494e70fee74b0f4bbf1351b/test/e2e/framework/size.go#L29-L61

> if a single test is importing bits from different c-provider SDK/API

Not so long ago that was exactly how it was done:
https://github.com/kubernetes/kubernetes/blob/def49db0585d31ea267d7677a24796b345b58e11/test/e2e/framework/size.go

Now the test goes through the ProviderInterface. While currently the
implementations of that are in-tree, they could be moved out.

> it feels as if such a test should be split into multiple tests to
> avoid the scenario.

That would lead to copy-and-paste of the test logic and thus make it
harder to update the tests later on. I think it is better to define a
test-specific interface and let cloud providers hook into that if they
want. Just to be clear, that interface doesn't have to be
ProviderInterface above.

> such tests should not reside in k/k.

Yes, a separate repo sounds better.

**Comment by elmiko** [comment]
fwiw, i am working on a plan to create/improve the cloud provider interface to the e2e tests, my hope is that once my initial investigation/poc is finished i will have a KEP to share about improving the tests for cloud specific functionality.

**Comment by elmiko** [comment]
there has been extensive work to remove cloud provider specific details from the e2e tests. i believe that currently only the fake provider is being used and we are no longer calling specific cloud functions to exercise the tests.

closing this issue for now, if we find more instances of cloud specific code we should open a new issue to target those instances.

/close
